### PR TITLE
ksw/implementation of descriptions and examples in the preference schema json

### PR DESCRIPTION
**Pull request by kswang1029**
This PR is to address https://github.com/CARTAvis/carta-frontend/issues/1339 and includes
1. a description per keyword
2. an example or enum when needed
3. revision of some input validation policies (correct me if the revision is not sensible)
4. add validation policy for color hex code.

Some descriptions contain ???, meaning I don't quite understand what the keyword means. Please advice.

@acdo2002 @YuHsuan-Hwang please help to check the descriptions and examples from a user point of view if you have time.

### CODE REVIEW

File: schemas/preference_schema_1.json [hunk 1]
```diff
@@ -9,64 +9,100 @@
...
             "type": "integer",
-            "minimum": 1
+            "minimum": 1,
+            "description": "Version number of the preference schema. An integer >= 1 need to be provided."
         },
         "username": {
             "type": "string",
-            "minLength": 1
+            "minLength": 1,
+            "description": "???"
```
Comment (veggiesaurus) [inline schemas/preference_schema_1.json:18]: best to just leave this description out. It's used by carta-controller

File: schemas/preference_schema_1.json [hunk 2]
```diff
@@ -9,64 +9,100 @@
...
         },
         "theme": {
             "type": "string",
+            "description": "Set the color theme of the graphical user interface of CARTA. Options are 'auto' (refering to OS theme if applicable), 'light', and 'dark'.",
```
Comment (veggiesaurus) [inline schemas/preference_schema_1.json:22]: do we need the options listed if we already have an enum?

File: schemas/preference_schema_1.json [hunk 3]
```diff
@@ -9,64 +9,100 @@
...
             "enum": [
                 "auto",
                 "light",
                 "dark"
             ]
         },
         "autoLaunch": {
-            "type": "boolean"
+            "type": "boolean",
```
Comment (veggiesaurus) [inline schemas/preference_schema_1.json:30]: if the default is true, we should probably put this in the schema as well using the `default` property

File: schemas/preference_schema_1.json (hunk 2 above)
Comment (kswang1029) [reply schemas/preference_schema_1.json:22]: I suggest we do. enum helps users to select one easily but if users want to revise what has been selected, the description helps I think.

File: schemas/preference_schema_1.json (hunk 1 above)
Comment (kswang1029) [reply schemas/preference_schema_1.json:18]: ok

File: schemas/preference_schema_1.json [hunk 4]
```diff
@@ -123,145 +191,249 @@
         "contourSmoothingMode": {
             "type": "integer",
             "minimum": 0,
-            "maximum": 2
+            "maximum": 2,
+            "description": "Set the image smoothing mode before computing contour vertices. An integer should be provided. Options are: {0: 'None', 1: 'Block', 2: 'Gaussian' (as default)}. Use 'contourSmoothingFactor' to define a kernel size for smoothing."
```
Comment (veggiesaurus) [inline schemas/preference_schema_1.json:195]: this is an example where we should switch to a string enum, I reckon. But that would be a breaking change, so we should combine it with other changes

File: schemas/preference_schema_1.json [hunk 5]
```diff
@@ -123,145 +191,249 @@
...
+            "minimum": 0,
+            "maximum": 50,
+            "description": "Set the default style of region rendering with dashed line in the image viewer. An integer between 0 (inclusive) and 50 (inclusive) should be given. Default is 0 (as solid line)."
         },
         "regionType": {
             "type": "integer",
             "minimum": 0,
-            "maximum": 6
+            "maximum": 6,
+            "description": "Set the default region type of the region creation button in the image viewer. An integer should be given. Options are: {0: 'Point', 1: 'Line' (NOT IMPLEMENTED YET), 2: 'Polyline' (NOT IMPLEMENTED YET), 3: 'Rectangle' (as default), 4: 'Ellipse', 5: 'Annulus' (NOT IMPLEMENTED YET), 6: 'Polygon'}."
```
Comment (veggiesaurus) [inline schemas/preference_schema_1.json:351]: we should make another enum string migration here

File: schemas/preference_schema_1.json [hunk 6]
```diff
@@ -9,64 +9,100 @@
...
                 "fuzzy",
                 "unix",
                 "regex"
             ]
         },
         "layout": {
-            "type": "string"
+            "type": "string",
+            "description": "Set a default layout when CARTA is initialized. Preset layouts are 'Default' (as default), 'Cube View', 'Cube Analysis', and 'Continuum Analysis'. A custom layout can be set as the default layout too.",
+            "enum": [
```
Comment (veggiesaurus) [inline schemas/preference_schema_1.json:64]: this shouldn't be an enum, because it can be any value. It should rather be a list of examples

Comment (kswang1029) [reply schemas/preference_schema_1.json:64]: uh true🙈 corrected

**Comment by kswang1029** [comment]
To generate an html manual, try (thanks to @veggiesaurus ):

> \# Install using "pip install json-schema-for-humans"
generate-schema-doc --minify --no-link-to-reused-ref --expand-buttons ../config/config_schema.json ../docs/config_schema.html

**Comment by kswang1029** [comment]
@veggiesaurus descriptions are revised and should be more concise now in the html manual. please re-review.

**Comment by acdo2002** [comment]
I checked `CARTAvis.github.io/schemas/preference_schema_1.json` description and found the corresponding function in the CARTA/File/Preference. 
At least for me, most of them are easy to understand what each function means.

(1) I do not know "fileSortingString" & "fileFilteringType"? which CARTA/File/Preference/Global function they correspond to?
(2) scaling description: allow integer 0-5, but scaling maximun: 7 , please check whether it is correct?
(3) Typo: Preference/Global/"Tranparent" (should be ''Transparent" right?) image background, I suppose this is a typo frontend issue?

**Comment by kswang1029** [comment]
@acdo2002 thanks for checking.
1) they are "implicit" preferences that need to be remembered by the frontend but there is no need to have them in the preferences menu.
2) that part will be fixed in other PR and will be changed to human-readable enum strings.
3) which part is it?

**Comment by acdo2002** [comment]
I see.
About (3) typo is on the carta frontend -> click "File" and click "Preference" menu bar -> check "Global". 
On the bottom, the last line shows
Tranparent Image Background
I suppose it should be Tran"s"parent Image Background? (now the screen shows "Tranparent" Image Background)

**Comment by kswang1029** [comment]
uh ok. I will fix it. 👍

**Comment by YuHsuan-Hwang** [comment]
I have read through the descriptions. Just found some ranges of the parameters different from UI, such as scalingGamma: [0, 2] and scalingAlpha: no limit. Do the ranges have to be consistent?

**Comment by kswang1029** [comment]
@YuHsuan-Hwang thanks for checking. Regarding your concerns, I suggest they should be consistent (that can be done in other PR along with two other human-readable enum changes. @veggiesaurus is that okay?).

**Comment by YuHsuan-Hwang** [comment]
Ok. Here I list some other params:
 - image/animation CompressionQuality: [1, 32]
 - contourSmoothingFactor: [1, 33]
 - contourNumLevels: [1, 20]

**Comment by kswang1029** [comment]
> Ok. Here I list some other params:
> 
> * image/animation CompressionQuality: [1, 32]
> * contourSmoothingFactor: [1, 33]
> * contourNumLevels: [1, 20]

Maybe I should get these value validation rules implemented. Will do so.

**Comment by kswang1029** [comment]
@YuHsuan-Hwang I have modified the value validators according to your findings. I manually set the upper bound for alpha to 1000000 as a sensible one.

**Review by veggiesaurus — APPROVED** [review]

File: schemas/preference_schema_1.json [hunk 7]
```diff
@@ -123,145 +212,282 @@
...
             "minimum": 128,
-            "maximum": 1024
+            "maximum": 1024,
+            "description": "ADJUST WITH CAUTION. Set the width of the contour control map for image matching. An integer between 128 and 1024 should be provided. The value needs to be multiple of 128. Default is 256.",
+            "default": 256
         },
         "streamContoursWhileZooming": {
-            "type": "boolean"
+            "type": "boolean",
+            "description": "should be 'streamImageTilesWhileZooming'???\n\nADJUST WITH CAUTION. A boolean flag to enable/disable image tile stream for intermediate zoom levels when image zooming is in action.",
```
Comment (confluence) [inline schemas/preference_schema_1.json:469]: This looks like an outstanding query about the preference name. Should we rename this preference in the code?

Comment (kswang1029) [reply schemas/preference_schema_1.json:469]: what it actually means is raster "image", not contour. So I think we better rename it to be consistent.

Comment (veggiesaurus) [reply schemas/preference_schema_1.json:469]: yes, we should probably rename it

**Comment by kswang1029** [comment]
ok to merge this?

**Review by YuHsuan-Hwang — APPROVED** [review]

**Review by acdo2002 — APPROVED** [review]

File: schemas/preference_schema_1.json (hunk 7 above)
Comment (confluence) [reply schemas/preference_schema_1.json:469]: This comment should be taken out before we merge. If the name needs to be changed, an issue should be made for it.

**Review by confluence — CHANGES_REQUESTED** [review]

File: schemas/preference_schema_1.json (hunk 7 above)
Comment (kswang1029) [reply schemas/preference_schema_1.json]: removed via 7280b64

**Review by confluence — APPROVED** [review]

**Pull request merged**